### DISCUSSION TITLE
Is there a way to handle multipart/form-data requests?

### DISCUSSION THREAD

**Discussion by Brandon-Perry** [trpc/trpc#658] [Q&A] [answered]
I would like to be able to upload images via a client.mutation, but I haven't been able to get it to work. I tried just adding a header with content-type set to multipart, as per below, but I just got back a 500 for that. Does TRPC support this? Or am I just doing this poorly?

```
const multipartClient = createTRPCClient<ApiRouter>({
    url: 'http://localhost:3000/v1',
    headers: {
        'content-type':'multipart/form-data'
    }
    
})
```

**Comment by KATT** [comment, marked as answer]
Hey @Brandon-Perry - no we don't support `multipart/form-data`. The way I've handled assets is to base64 encode and pass as part as part of JSON-payload. Not ideal, but it works quite well.

**Reply by connordotlol** [reply]
For large uploads, something like [formidable](https://www.npmjs.com/package/formidable) will store uploaded files to a temporary location on disk so that the server doesn't need to store the entire upload in memory all at one time. The same would be true for the client, which wouldn't need to load a base64 encoded string of the file data into memory either. (I think? Correct me if I'm wrong on that.)

I'm working on a project that requires file uploads, and the files my application takes can be quite large; it's feasible that they'd be too large to fit into memory at some point in the future. I love the concept of trpc, but in order to use it in my project, I'm going to have to write some kind of middleware to handle file uploads separately from trpc due to the memory concerns. Is multipart handling something that the trpc team is interested in implementing in the future?

**Reply by KATT** [reply]
Hey @connordotlol - the tRPC team is largely a one-man band (me 👋) and I don't have any plans on making tRPC anything else but JSON-only in the near future.

There are ways around it by simply making an API-handler that handles the upload, which is dead-easy in Next.js. I personally use S3 for large objects - I do this by having a tRPC-endpoint that returns a pre-signed URL for doing the upload straight to S3.

**Reply by connordotlol** [reply]
Thanks! Sorry about the "team" thing, guess that's a habit. I'm avoiding becoming dependent on opaque third party dependencies like Next and S3, but the separate route handler definitely isn't a hassle when you take into account all the time saved by trpc. Thank you for the work you do.

**Reply by KATT** [reply]
Thanks for the kind words! 🙏

**Reply by juliozengcheil** [reply]
@KATT hey thanks for the reply but the issue i have, is the isSubmitting, i use it to show the loading icon but it need to upload 4 files and then call the mutate function and i think thats change isSubmitting to true, is there a way to do it cleanly? or better?

**Comment by KATT** [comment]
See #1937
